File: src/main/scala/CmdRegistry.scala
package codecraft.codegen

import util.Try

final case class GroupRouting(
  queueName: String,
  routingKey: String,
  exchange: String
)

final case class CmdRegistry(
  key: String,
  serializeCmd: Any => Try[Array[Byte]],
  serializeReply: Any => Try[Array[Byte]],
  deserializeCmd: Array[Byte] => Try[Any],
  deserializeReply: Array[Byte] => Try[Any],
  group: GroupRouting
)

File: src/main/fsharp/CmdRegistry.fs
namespace codecraft.codegen

type 'a Try = Choice<'a, exn>

type GroupRouting = {
  queueName: string
  routingKey: string
  exchange: string
}

type CmdRegistry = {
  key: string
  serializeCmd: obj -> byte array Try
  serializeReply: obj -> byte array Try
  deserializeCmd: byte array -> obj Try
  deserializeReply: byte array -> obj Try
  group: GroupRouting
}

type EventRegistry = {
  key: string
  serialize: obj -> byte array Try
  deserialize: byte array -> obj Try
  group: GroupRouting
}
